pkg/reconciler/ingress/ingress.go

```diff
@@ -148,13 +150,30 @@ func (r *Reconciler) reconcileIngress(ctx context.Context, ing *v1alpha1.Ingress
 			if err != nil {
 				return err
 			}
-			desired, err := resources.MakeTLSServers(ing, ns, originSecrets)
+
+			nonWildcardIngressTLS := resources.GetNonWildcardIngressTLS(ing.Spec.TLS, nonWildcardSecrets)
+			// For Ingress TLS referencing non-wildcard certificate, we reconcile user provided Gateway to enforce the TLS configuration regularly.
+			desiredIngressServer, err := resources.MakeTLSServers(ing, nonWildcardIngressTLS, ns, nonWildcardSecrets)
 			if err != nil {
 				return err
 			}
-			if err := r.reconcileIngressServers(ctx, ing, gw, desired); err != nil {
+			if err := r.reconcileIngressServers(ctx, ing, gw, desiredIngressServer); err != nil {
 				return err
 			}
+
+			// For Ingress TLS referencing wildcard certificates, we reconcile a separate Gateway
+			// that will be shared by other Ingresses that reference the
+			// same wildcard host. We need to handle wildcard certificate specially because Istio does
+			// not fully support multiple TLS Servers (or Gateways) share the same certificate.
+			// https://istio.io/docs/ops/common-problems/network-issues/
+			desiredWildcardGateways, err := resources.MakeWildcardGateways(ctx, wildcardSecrets, r.svcLister)
+			if err != nil {
+				return err
+			}
+			if err := r.reconcileWildcardGateways(ctx, desiredWildcardGateways); err != nil {
```

Comment:
This is a code in the ingres reconcile process. So if only Gateway is manually updated, this function is not called and it is not reconciled.
I mean fox example - gateway is not reonciled until Kingres is refreshed:

```
$ kubectl get gateway -n bar
NAME               AGE
wildcard-df913fc   23s

$ kubectl delete gateway -n bar --all
gateway.networking.istio.io "wildcard-df913fc" deleted

$ kubectl get gateway -n bar
No resources found.

# Gateway is not created until one of Kingress is refreshed or new one is created.
```

Reply:
But it is same with non-wildcard gateway, so I wonder if we should fix it or not.

Reply:
Good question.
For the global gateway under knative-serving namespace, the Gateway change won't trigger Ingress reconcilation because it may cause the situation that one Ingress updates the Gateway which cause the reconcilation of all Ingresses, which is a kind of expensive. In addtion, I will work on migrate the global Gateway model to one-Gateway-per-Ingress model. Once we have this model, the change of a Gateway belonging to one Ingress will trigger the reconcilation of that Ingress.

For the wildcard Gateway, I think it makes sense to trigger the reconcilation of Ingresses when the Gateway is changed or deleted as it won't have the situation of the global Gateway. I added tracker to track the Ingresses associated with the wildcard Gateway. Now manually deleted Gateway will be recreated.